src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePatternAnalyzer.cs

```diff
@@ -31,94 +31,96 @@ private void AnalyzeSemanticModel(SemanticModelAnalysisContext context)
         var cancellationToken = context.CancellationToken;
 
         var root = syntaxTree.GetRoot(cancellationToken);
-        var wellKnownTypes = WellKnownTypes.GetOrCreate(context.SemanticModel.Compilation);
         var routeUsageCache = RouteUsageCache.GetOrCreate(context.SemanticModel.Compilation);
-        Analyze(context, root, wellKnownTypes, routeUsageCache, cancellationToken);
-    }
 
-    private void Analyze(
-        SemanticModelAnalysisContext context,
-        SyntaxNode node,
-        WellKnownTypes wellKnownTypes,
-        RouteUsageCache routeUsageCache,
-        CancellationToken cancellationToken)
-    {
-        cancellationToken.ThrowIfCancellationRequested();
+        // Use a stack collection so that we don't blow the actual stack through recursion.
+        var stack = new Stack<SyntaxNode>();
+        stack.Push(root);
 
-        foreach (var child in node.ChildNodesAndTokens())
+        while (stack.Count != 0)
         {
-            if (child.IsNode)
-            {
-                Analyze(context, child.AsNode()!, wellKnownTypes, routeUsageCache, cancellationToken);
-            }
-            else
+            cancellationToken.ThrowIfCancellationRequested();
+            var current = stack.Pop();
+
+            foreach (var child in current.ChildNodesAndTokens())
```

Comment:
>foreach (var child in current.ChildNodesAndTokens()) [](http://example.com/codeflow?start=12&length=52)

It looks like the analysis is only interested in tokens, so the loops could be replaced with `foreach (var token in root.DescendantTokens()) { ... }`.

Reply:
The code I added follows what the Regex analyzer does. I investigated why the regex analyzer doesn't use `DecendantTokens`. I think it could have, but the recently added `FilteredSpans` check wouldn't work with the helper method:

https://github.com/dotnet/roslyn/blob/4dcbb0bdb9f7630fec6ce81b311525ef974fd4b0/src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/AbstractRegexDiagnosticAnalyzer.cs#L63-L64

It looks like `FilteredSpans` has only just been made public. aspnetcore probably won't update to bits with it until after .NET 8. I'd like to leave the method as is so using FilteredSpans in the future is possible.

Btw, what are filtered spans? I'm guessing it's an optimization but I'd be interested if you could point me at an example of it.

Reply:
Adding @mavasani for the question on filtered spans.

Reply:
FilterSpan is the new API added on analysis context types when computing diagnostics for lightbulb. This allows expensive analyzers to optimize their performance based on the span within the tree for which diagnostics are being computed. If the analyzer ignores this span, the analyzer driver will filter the reported diagnostics outside this span, there should be no functional impact of ignoring this filter span.

Reply:
This analyzer is expensive. Do you recommend eventually using filter span [like other analyzers are doing](https://github.com/dotnet/roslyn/blob/4dcbb0bdb9f7630fec6ce81b311525ef974fd4b0/src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/AbstractRegexDiagnosticAnalyzer.cs#L63-L64)?

Reply:
Definitely - all SemanticModelAnalysisContext callbacks should respect the filter span. I plan to write a meta-analyzer in Microsoft.CodeAnalysis.Analyzers package to recommend the same. These analyzers are known to be the biggest beneficiaries of this API on lightbulb path.

Reply:
I do not understand why `DescendantTokens` is in compatible with `FilteredSpans`. It can be implemented by passing down the predicate correct? Having every analyzer / generator out there build their own stack for navigating the tree to scan tokens is **not** the right solution.

@333fred

Reply:
I see there is a `DescendantTokens` overload that takes a span: https://github.com/dotnet/roslyn/blob/252def9efa028af3f0562929b4b0e8a74dcd3124/src/Compilers/Core/Portable/Syntax/SyntaxNode.cs#L1049

I looks like checks content overlaps.

I think this would work (assuming FilterSpan was available):

```cs
foreach (var item in root.DescendantTokens(context.FilterSpan))
{
    cancellationToken.ThrowIfCancellationRequested();

    AnalyzeToken(context, routeUsageCache, item, cancellationToken);
}
```